File: src/ib/ib-client.ts
import { IBApi, EventName, Contract, BarSizeSetting } from "@stoqey/ib";

export class IBClient {
  private ib: IBApi;
  private connected = false;
  private ready = false;
  private reqId = 1;

  constructor(
    private readonly host: string,
    private readonly port: number,
    private readonly clientId: number
  ) {
    this.ib = new IBApi({ host: this.host, port: this.port, clientId: this.clientId, timeout: 1000  });

    this.ib.on(EventName.connected, () => {
      this.connected = true;
      console.log(JSON.stringify({
        level: "info",
        msg: "IB socket connected",
        ts: new Date().toISOString(),
        host: this.host,
        port: this.port,
        clientId: this.clientId,
      }));

      // 1=Live, 2=Frozen, 3=Delayed, 4=Delayed-Frozen
      this.ib.reqMarketDataType(2);

      // Nudge the gateway to emit *something* we can call "ready"
      try { this.ib.reqIds(1); } catch {}
      try { /* some builds expose this */ // @ts-ignore
        this.ib.reqManagedAccts?.();
      } catch {}
      try { this.ib.reqCurrentTime(); } catch {}
    });

    this.ib.on(EventName.disconnected, () => {
      this.connected = false;
      this.ready = false;
      console.warn(JSON.stringify({
        level: "warn",
        msg: "IB socket disconnected; retrying in 5s",
        ts: new Date().toISOString()
      }));
      setTimeout(() => this.connect().catch(() => void 0), 5000);
    });

    // Mark "ready" if any of these arrive (but we no longer hard-require them)
    const markReady = (source: string) => {
      if (!this.ready) {
        this.ready = true;
        console.log(JSON.stringify({
          level: "info",
          msg: "IB ready signal received",
          source,
          ts: new Date().toISOString()
        }));
      }
    };
    this.ib.on(EventName.managedAccounts, () => markReady("managedAccounts"));
    this.ib.on(EventName.nextValidId, () => markReady("nextValidId"));
    this.ib.on(EventName.currentTime, () => markReady("currentTime"));

    this.ib.on(EventName.error, (err) => {
      console.error(JSON.stringify({
        level: "error",
        msg: "IB error",
        ts: new Date().toISOString(),
        error: (err as any)?.message || String(err)
      }));
    });
  }

  /** Connect and wait for TCP connect (not full handshake). */
  public async connect(): Promise<void> {
    if (this.connected) return;

    console.log(JSON.stringify({
      level: "info",
      msg: "Connecting to IB socket…",
      ts: new Date().toISOString(),
      host: this.host,
      port: this.port,
      clientId: this.clientId
    }));

    this.ib.connect(this.clientId);
    await this.waitForConnected(12_000);

    // Kick nudges again after TCP connect
    try { this.ib.reqIds(1); } catch {}
    try { /* @ts-ignore */ this.ib.reqManagedAccts?.(); } catch {}
    try { this.ib.reqCurrentTime(); } catch {}
  }

  public isConnected(): boolean {
    return this.connected;
  }

/** Quick TCP/connectivity probe without touching the main client session. */
public async testConnection(timeoutMs = 5000): Promise<{ connected: boolean; error?: string }> {
  const probeClientId = this.clientId + 10; // avoid clashing with main session
  const probe = new IBApi({ host: this.host, port: this.port, clientId: probeClientId });

  return new Promise<{ connected: boolean; error?: string }>((resolve) => {
    let done = false;
    const finish = (ok: boolean, error?: string) => {
      if (done) return;
      done = true;
      try { probe.disconnect(); } catch {}
      resolve(ok ? { connected: true } : { connected: false, error });
    };

    const timer = setTimeout(() => finish(false, `Connection timeout after ${timeoutMs}ms`), timeoutMs);

    probe.on(EventName.connected, () => {
        console.log("HELLO");
      clearTimeout(timer);
      // Nudge to ensure API is actually responsive (optional)
      try { probe.reqCurrentTime(); } catch {}
      // If you want a stronger check, wait for currentTime once with a short timeout
      let ctTimer: any = setTimeout(() => finish(true), 500); // soft success if no reply
      probe.once(EventName.currentTime, () => {
        clearTimeout(ctTimer);
        finish(true);
      });
    });

    probe.on(EventName.error, (err: any) => {
      clearTimeout(timer);
      console.log(err)
      finish(false, err?.message || String(err));
    });

    try {
      probe.connect(probeClientId);
    } catch (e: any) {
      clearTimeout(timer);
      finish(false, e?.message || String(e));
    }
  });
}

  /** Wait until TCP connected or throw. */
  public async waitForConnected(timeoutMs = 10_000): Promise<void> {
    if (this.connected) return;
    await new Promise<void>((resolve, reject) => {
      const t = setTimeout(() => {
        reject(new Error("IB TCP connect timeout"));
      }, timeoutMs);

      const onConn = () => { clearTimeout(t); cleanup(); resolve(); };
      const onErr = (e: any) => { clearTimeout(t); cleanup(); reject(e instanceof Error ? e : new Error(String(e))); };
      const cleanup = () => {
        this.ib.off(EventName.connected, onConn);
        this.ib.off(EventName.error, onErr);
      };

      this.ib.on(EventName.connected, onConn);
      this.ib.on(EventName.error, onErr);
    });
  }


  private nextReqId(): number {
    this.reqId = (this.reqId + 1) % 2147480000;
    if (this.reqId <= 0) this.reqId = 1;
    return this.reqId;
  }

  public async reqContractDetails(contract: Partial<Contract>, timeoutMs = 12_000) {
    await this.waitForConnected();
    const reqId = this.nextReqId();

    return new Promise<any[]>((resolve, reject) => {
      const details: any[] = [];
      const t0 = Number(process.hrtime.bigint() / 1_000_000n);

      const timer = setTimeout(() => {
        cleanup();
        reject(new Error(`contractDetails timeout after ${timeoutMs}ms`));
      }, timeoutMs);

      const onDetails = (id: number, d: any) => { if (id === reqId) details.push(d); };
      const onEnd = (id: number) => {
        if (id !== reqId) return;
        cleanup();
        const ms = Number(process.hrtime.bigint() / 1_000_000n) - t0;
        console.log(JSON.stringify({
          level: "info",
          msg: "contractDetails done",
          ms,
          count: details.length,
          first: pickContractFields(details?.[0]?.contract),
          ts: new Date().toISOString(),
        }));
        resolve(details);
      };
      const onErr = (err: any) => { cleanup(); reject(err instanceof Error ? err : new Error(JSON.stringify(err))); };
      const cleanup = () => {
        clearTimeout(timer);
        this.ib.off(EventName.contractDetails, onDetails);
        this.ib.off(EventName.contractDetailsEnd, onEnd);
        this.ib.off(EventName.error, onErr);
      };

      this.ib.on(EventName.contractDetails, onDetails);
      this.ib.on(EventName.contractDetailsEnd, onEnd);
      this.ib.on(EventName.error, onErr);

      console.log(JSON.stringify({
        level: "info",
        msg: "contractDetails request",
        reqId,
        base: pickContractFields(contract as any),
        ts: new Date().toISOString(),
      }));

      this.ib.reqContractDetails(reqId, contract as Contract);
    });
  }

  

  public async reqHistoricalData(opts: {
    contract: Contract;
    endDateTime?: string;
    durationStr: string;
    barSize: BarSizeSetting;
    whatToShow?: "TRADES" | "MIDPOINT" | "BID" | "ASK" | "BID_ASK" | "ADJUSTED_LAST";
    useRTH?: 0 | 1;
    formatDate?: 1 | 2;
    keepUpToDate?: boolean;
    timeoutMs?: number;
  }) {
    await this.waitForConnected();
    const {
      contract,
      endDateTime = "",
      durationStr,
      barSize,
      whatToShow = "TRADES",
      useRTH = 0,
      formatDate = 1,
      keepUpToDate = false,
      timeoutMs = 30_000,
    } = opts;

    const reqId = this.nextReqId();

    return new Promise<Array<{
      time: string;
      open: number; high: number; low: number; close: number;
      volume: number; count?: number; WAP: number; hasGaps?: boolean;
    }>>((resolve, reject) => {
      const rows: any[] = [];
      const t0 = Number(process.hrtime.bigint() / 1_000_000n);
      let hardTimer: any;
      let idleTimer: any;

      const clearTimers = () => {
        if (hardTimer) clearTimeout(hardTimer);
        if (idleTimer) clearTimeout(idleTimer);
        hardTimer = idleTimer = undefined;
      };

      const scheduleIdleResolve = () => {
        if (idleTimer) clearTimeout(idleTimer);
        idleTimer = setTimeout(() => {
          cleanup();
          const ms = Number(process.hrtime.bigint() / 1_000_000n) - t0;
          console.log(JSON.stringify({
            level: "info",
            msg: "historicalData idle-resolve",
            reqId, ms, count: rows.length,
            firstTime: rows[0]?.time, lastTime: rows[rows.length - 1]?.time,
            ts: new Date().toISOString(),
          }));
          resolve(rows);
        }, 800);
      };

      hardTimer = setTimeout(() => {
        cleanup();
        reject(new Error(`historicalData timeout after ${timeoutMs}ms`));
      }, timeoutMs);

      const onRow = (
        id: number,
        time: string,
        open: number,
        high: number,
        low: number,
        close: number,
        volume: number,
        count: number | undefined,
        WAP: number,
        hasGaps: boolean | undefined
      ) => {
        if (id !== reqId) return;

        if (typeof time === "string" && time.startsWith("finished-")) {
          cleanup();
          const ms = Number(process.hrtime.bigint() / 1_000_000n) - t0;
          console.log(JSON.stringify({
            level: "info",
            msg: "historicalData finished-marker",
            reqId, ms, count: rows.length,
            firstTime: rows[0]?.time, lastTime: rows[rows.length - 1]?.time,
            ts: new Date().toISOString(),
          }));
          return resolve(rows);
        }

        if (open === -1 && close === -1) return;

        rows.push({ time, open, high, low, close, volume, count, WAP, hasGaps });
        scheduleIdleResolve();
      };

      const onEnd = (id: number) => {
        if (id !== reqId) return;
        cleanup();
        const ms = Number(process.hrtime.bigint() / 1_000_000n) - t0;
        console.log(JSON.stringify({
          level: "info",
          msg: "historicalData end",
          reqId, ms, count: rows.length,
          firstTime: rows[0]?.time, lastTime: rows[rows.length - 1]?.time,
          ts: new Date().toISOString(),
        }));
        resolve(rows);
      };

      const onErr = (err: any) => {
        cleanup();
        console.error(JSON.stringify({
          level: "error",
          msg: "historicalData error",
          reqId,
          error: err instanceof Error ? err.message : JSON.stringify(err),
          ts: new Date().toISOString(),
        }));
        reject(err instanceof Error ? err : new Error(JSON.stringify(err)));
      };

      const cleanup = () => {
        clearTimers();
        this.ib.off(EventName.historicalData, onRow);
        // @ts-ignore may not exist
        this.ib.off?.(EventName.historicalDataEnd as any, onEnd);
        this.ib.off(EventName.error, onErr);
      };

      console.log(JSON.stringify({
        level: "info",
        msg: "historicalData request",
        ts: new Date().toISOString(),
        reqId,
        params: { endDateTime, durationStr, barSize, whatToShow, useRTH, formatDate, keepUpToDate },
        contract: pickContractFields(contract),
      }));

      this.ib.on(EventName.historicalData, onRow);
      // @ts-ignore may not exist
      this.ib.on?.(EventName.historicalDataEnd as any, onEnd);
      this.ib.on(EventName.error, onErr);

      this.ib.reqHistoricalData(
        reqId,
        contract,
        endDateTime,
        durationStr,
        barSize,
        whatToShow,
        useRTH,
        formatDate,
        keepUpToDate
      );
    });
  }
}

function pickContractFields(c: Partial<Contract> | undefined) {
  if (!c) return c;
  const { symbol, secType, exchange, currency, localSymbol, conId, lastTradeDateOrContractMonth } = c as any;
  return { symbol, secType, exchange, currency, localSymbol, conId, lastTradeDateOrContractMonth };
}

File: src/ib/ib_client.py
import asyncio
import logging
import time
import random
from datetime import datetime
from typing import List, Optional, Dict, Any
from ibapi.client import EClient
from ibapi.wrapper import EWrapper
from ibapi.contract import Contract
from ibapi.common import BarData
from threading import Thread, Event
import queue

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class IBClient(EWrapper, EClient):
    def __init__(self, host: str = "127.0.0.1", port: int = 7497, client_id: int = None, timeout: int = 10):
        EClient.__init__(self, self)
        EWrapper.__init__(self)
        
        self.host = host
        self.port = port
        self.client_id = client_id if client_id is not None else random.randint(1, 10000)
        self.timeout = timeout
        self.connected = False
        self.ready = False
        self.next_order_id = None
        
        # Request ID management
        self.req_id = 1
        
        # Data storage for requests
        self.contract_details = {}
        self.historical_data = {}
        self.request_events = {}
        
    def nextValidId(self, orderId: int):
        """Called when connection is established"""
        self.next_order_id = orderId
        self.ready = True
        logger.info(f"Connected to IB. Next valid order ID: {orderId}")
        
    def connectAck(self):
        """Called when connection is acknowledged"""
        self.connected = True
        logger.info("IB connection acknowledged")
        
    def connectionClosed(self):
        """Called when connection is closed"""
        self.connected = False
        self.ready = False
        logger.warning("IB connection closed")
        
    def error(self, reqId: int, errorCode: int, errorString: str, advancedOrderRejectJson: str = ""):
        """Handle errors"""
        logger.error(f"Error {errorCode}: {errorString} (reqId: {reqId})")
        
        # Signal error for pending requests
        if reqId in self.request_events:
            self.request_events[reqId].set()
            
    def contractDetails(self, reqId: int, contractDetails):
        """Receive contract details"""
        if reqId not in self.contract_details:
            self.contract_details[reqId] = []
        self.contract_details[reqId].append(contractDetails)
        
    def contractDetailsEnd(self, reqId: int):
        """Contract details request completed"""
        logger.info(f"Contract details completed for reqId {reqId}")
        if reqId in self.request_events:
            self.request_events[reqId].set()
            
    def historicalData(self, reqId: int, bar: BarData):
        """Receive historical data"""
        if reqId not in self.historical_data:
            self.historical_data[reqId] = []
            
        self.historical_data[reqId].append({
            'time': bar.date,
            'open': bar.open,
            'high': bar.high,
            'low': bar.low,
            'close': bar.close,
            'volume': bar.volume,
            'count': bar.barCount,
            'wap': bar.wap
        })
        
    def historicalDataEnd(self, reqId: int, start: str, end: str):
        """Historical data request completed"""
        logger.info(f"Historical data completed for reqId {reqId}")
        if reqId in self.request_events:
            self.request_events[reqId].set()
    
    def connect_to_ib(self, timeout: int = None) -> bool:
        """Connect to Interactive Brokers"""
        connection_timeout = timeout if timeout is not None else self.timeout
        
        try:
            logger.info(f"Connecting to IB at {self.host}:{self.port} with client ID {self.client_id} (timeout: {connection_timeout}s)")
            self.connect(self.host, self.port, self.client_id)
            
            # Start message processing thread
            thread = Thread(target=self.run)
            thread.daemon = True
            thread.start()
            
            # Wait for connection
            start_time = time.time()
            while not self.ready and (time.time() - start_time) < connection_timeout:
                time.sleep(0.1)
                
            if self.ready:
                logger.info("Successfully connected to IB")
                return True
            else:
                logger.error(f"Failed to connect to IB within {connection_timeout}s timeout")
                return False
                
        except Exception as e:
            logger.error(f"Connection error: {e}")
            return False
    
    def disconnect_from_ib(self):
        """Disconnect from Interactive Brokers"""
        self.disconnect()
        self.connected = False
        self.ready = False
        
    def get_next_req_id(self) -> int:
        """Get next request ID"""
        req_id = self.req_id
        self.req_id += 1
        return req_id
        
    def test_connection(self, timeout: int = 5) -> Dict[str, Any]:
        """Test connection to IB"""
        try:
            # Try to connect temporarily with random client ID
            test_client_id = random.randint(10000, 99999)
            test_client = IBClient(self.host, self.port, test_client_id, timeout)
            if test_client.connect_to_ib(timeout):
                test_client.disconnect_from_ib()
                return {"connected": True}
            else:
                return {"connected": False, "error": "Connection timeout"}
        except Exception as e:
            return {"connected": False, "error": str(e)}
    
    def req_contract_details(self, contract: Contract, timeout: int = 10) -> List[Any]:
        """Request contract details"""
        if not self.ready:
            raise Exception("Not connected to IB")
            
        req_id = self.get_next_req_id()
        self.request_events[req_id] = Event()
        
        logger.info(f"Requesting contract details for {contract.symbol} (reqId: {req_id})")
        self.reqContractDetails(req_id, contract)
        
        # Wait for response
        if self.request_events[req_id].wait(timeout):
            details = self.contract_details.get(req_id, [])
            # Cleanup
            del self.request_events[req_id]
            if req_id in self.contract_details:
                del self.contract_details[req_id]
            return details
        else:
            # Cleanup on timeout
            del self.request_events[req_id]
            if req_id in self.contract_details:
                del self.contract_details[req_id]
            raise Exception(f"Contract details request timeout for reqId {req_id}")
    
    def req_historical_data(self, contract: Contract, end_date: str = "", 
                          duration: str = "1 D", bar_size: str = "1 hour",
                          what_to_show: str = "TRADES", use_rth: int = 0,
                          timeout: int = 30) -> List[Dict[str, Any]]:
        """Request historical data"""
        if not self.ready:
            raise Exception("Not connected to IB")
            
        req_id = self.get_next_req_id()
        self.request_events[req_id] = Event()
        
        logger.info(f"Requesting historical data for {contract.symbol} (reqId: {req_id})")
        self.reqHistoricalData(
            req_id, contract, end_date, duration, bar_size,
            what_to_show, use_rth, 1, False, []
        )
        
        # Wait for response
        if self.request_events[req_id].wait(timeout):
            data = self.historical_data.get(req_id, [])
            # Cleanup
            del self.request_events[req_id]
            if req_id in self.historical_data:
                del self.historical_data[req_id]
            return data
        else:
            # Cleanup on timeout
            del self.request_events[req_id]
            if req_id in self.historical_data:
                del self.historical_data[req_id]
            raise Exception(f"Historical data request timeout for reqId {req_id}")

def create_contract(symbol: str, sec_type: str = "STK", exchange: str = "SMART", currency: str = "USD") -> Contract:
    """Create a contract object"""
    contract = Contract()
    contract.symbol = symbol
    contract.secType = sec_type
    contract.exchange = exchange
    contract.currency = currency
    return contract